3scaleAdapter/pkg/threescale/threescale.go

```diff
@@ -119,44 +235,48 @@ func (s *Threescale) isAuthorized(cfg *config.Params, request authorization.Inst
 
 	if proxyConfErr != nil {
 		return status.WithMessage(
-			9, fmt.Sprintf("currently unable to fetch required data from 3scale system - %s", proxyConfErr.Error())), err
+			9, fmt.Sprintf("currently unable to fetch required data from 3scale system - %s", proxyConfErr.Error())), nil
 	}
 
-	authType := pce.ProxyConfig.Content.BackendAuthenticationType
-	switch authType {
-	case "provider_key", "service_token":
-		return s.doAuthRep(cfg.ServiceId, userKey, request, pce.ProxyConfig)
-	default:
-		err := fmt.Errorf("unsupported auth type %s for service %s", authType, cfg.ServiceId)
-		return status.WithMessage(16, err.Error()), err
-
-	}
+	return s.doAuth(cfg.ServiceId, userKey, request, pce.ProxyConfig)
 }
-
-func (s *Threescale) doAuthRep(svcID string, userKey string, request authorization.InstanceMsg, conf sysC.ProxyConfig) (rpc.Status, error) {
+func (s *Threescale) doAuth(svcID string, userKey string, request authorization.InstanceMsg, conf sysC.ProxyConfig) (rpc.Status, error) {
 	var resp backendC.ApiResponse
 	var apiErr error
 
+	if request.Action.Path == "" {
+		return status.WithInvalidArgument("missing request path"), nil
+	}
+
 	bc, err := s.backendClientBuilder(conf.Content.Proxy.Backend.Endpoint)
 	if err != nil {
 		return status.WithInvalidArgument(
 			fmt.Sprintf("error creating 3scale backend client - %s", err.Error())), err
 	}
 
-	shouldReport, params := s.generateReports(request, conf)
-	if !shouldReport {
+	metrics := generateMetrics(request.Action.Path, request.Action.Method, conf)
+	if len(metrics) == 0 {
 		return status.WithPermissionDenied(fmt.Sprintf("no matching mapping rule for request with method %s and path %s",
 			request.Action.Method, request.Action.Path)), nil
 	}
 
-	if conf.Content.BackendAuthenticationType == "provider_key" {
-		resp, apiErr = bc.AuthRepProviderKey(userKey, conf.Content.BackendAuthenticationValue, svcID, params)
-	} else {
-		resp, apiErr = bc.AuthRepKey(userKey, conf.Content.BackendAuthenticationValue, svcID, params)
+	switch conf.Content.BackendAuthenticationType {
+
+	case "provider_key":
+		params := backendC.NewAuthorizeKeyParams("", "")
+		params.Metrics = metrics
+		resp, apiErr = bc.AuthorizeKey(userKey, conf.Content.BackendAuthenticationValue, svcID, params)
+	case "service_token":
+		params := backendC.NewAuthorizeKeyParams("", "")
+		params.Metrics = metrics
+		resp, apiErr = bc.AuthorizeKey(userKey, conf.Content.BackendAuthenticationValue, svcID, params)
```

Comment:
Unless I am mistaken the two blocks of code above are doing the same thing? In which case you can have a single case `"provider_key", "service_token"`and a catch all rest in default